### ISSUE TITLE
Delay memoization cache hit check until actual step execution

### ISSUE THREAD

**Issue by avasiliev** [argoproj/argo-workflows#10109]
# Summary

Don't check memoization cache hits on workflow startup, because delaying this check may save some computational resources.

# Use Cases

I would like to save some cluster resources not only when some work is already done and the artifacts are ready, but even when multiple jobs are trying to do the same task simultaneously.

Here is a simplified example:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: memoized-workflow-test-
spec:
  entrypoint: main
  templates:
  - name: main
    steps:
    - - name: run-1
        template: run
        arguments:
          parameters:
          - name: id
            value: 1
          - name: cache-key
            value: the-same-in-all-runs
      - name: run-2
        template: run
        arguments:
          parameters:
          - name: id
            value: 2
          - name: cache-key
            value: the-same-in-all-runs
  - name: run
    inputs:
      parameters:
      - name: id
      - name: cache-key
    script:
      image: ubuntu:22.04
      command: [bash]
      source: |
        sleep 30
        echo result: {{inputs.parameters.id}}
    memoize:
      key: "{{inputs.parameters.cache-key}}"
      cache:
        configMap:
          name: some-test-cache
    synchronization:
      mutex:
        name: 'memoized-workflow-test-run-{{inputs.parameters.cache-key}}'
```

Here I have two steps which do the same job (and that is reflected in the fact that they have the same cache key). These steps in theory may run in parallel, but I introduced the mutex, which is unique for each possible cache key. 

As a result I would expect that actual useful job is done exactly once: by the step which managed to acquire the mutex first. But unfortunately, the status of cache hit is checked only at workflow startup, so the job is done twice and some computational resources are wasted.

That's a simplified example, and of course this semantics is [also] needed to save resources across different workflows.

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

**Comment by terrytangyuan** [comment]
This is a great idea. Would you like to look into implementing this?

**Comment by avasiliev** [comment]
Thanks. We may consider this, but have no idea where to start. Maybe someone can provide any hints?

**Comment by terrytangyuan** [comment]
The main logic is in https://github.com/argoproj/argo-workflows/blob/master/workflow/controller/operator.go

Search for "memoize".

**Comment by Joibel** [comment]
This has since been implemented. The memoization cache is no longer only consulted when the node is created: it is re-checked after a node acquires its synchronization lock, so in your example job-2 will pick up job-1's cache entry when it gets the mutex and will not re-run the work.

Fixed in PR #11456 (released in v3.5.0, backported to v3.4.10 as #11578), with a follow-up for concurrent workflows in #11883. The behaviour is covered by TestWorkflowMemoizationWithMutex, which is close to the workflow you posted.

Closing as already fixed - please reopen or file a new issue if you still see duplicated work on a currently supported version.
